### PR TITLE
Support bzlmod

### PR DESCRIPTION
**Pull request by cameron-martin**
This makes rules_rust load dependencies via bzlmod. Currently only basic functionality is completed, such as registering rustc toolchains and compiling crates. Note that it cannot interact with cargo, wasm or load any other rust toolchains such as rustfmt.
    
There is one new module, `examples/bzlmod/hello_world`, that depends on the root `rules_rust` module. This example can be built and run using:
    
```
cd examples/bzlmod/hello_world
bazel run //:hello_world
```

To register toolchains in an ergonomic way, it defines a new "hub" repository that contains all the toolchain proxies, so they can all be registered like so:

```
register_toolchains("@rust_toolchains//:all")
```

closes https://github.com/bazelbuild/rules_rust/issues/1493

### CODE REVIEW

**Comment by UebelAndre** [comment]
re https://github.com/bazelbuild/rules_rust/issues/1493#issuecomment-1227590691

@scentini do you or anyone you have contact with at Google know anything about bzlmod?

**Comment by scentini** [comment]
I'm not familiar with bzlmod, in case we're doing something silly here it's not obvious to me. @Wyverald could you PTAL?

**Comment by scentini** [comment]
cc @meteorcloudy

**Comment by meteorcloudy** [comment]
Bzlmod may have some bugs in Bazel 5.3.0, it's recommended to test with Bazel@HEAD, you can use Bazelisk:
```
export USE_BAZEL_VERSION=last_green
bazelisk build //:hello_world
```
I got the following toolchain error instead of the visibility error:
```
ERROR: /Users/pcloudy/workspace/rules_rust/examples/bzlmod/hello_world/BUILD.bazel:5:12: While resolving toolchains for target //:hello_world: No matching toolchains found for types @rules_rust~override//rust:toolchain_type.
To debug, rerun with --toolchain_resolution_debug='@rules_rust~override//rust:toolchain_type'
If platforms or toolchains are a new concept for you, we'd encourage reading https://bazel.build/concepts/platforms-intro.
```

File: rust/private/repository_utils.bzl [hunk 1]
```diff
@@ -205,15 +205,15 @@ load("@rules_rust//rust:toolchain.bzl", "rust_toolchain")
 
 rust_toolchain(
     name = "{toolchain_name}",
-    rust_doc = "@{workspace_name}//:rustdoc",
-    rust_std = "@{workspace_name}//:rust_std-{target_triple}",
-    rustc = "@{workspace_name}//:rustc",
+    rust_doc = "//:rustdoc",
```
Comment (cameron-martin) [inline rust/private/repository_utils.bzl:208]: If I try to use the repository name in this, I get the following error:

```
ERROR: /home/cameron/.cache/bazel/_bazel_cameron/538574de729809a828a75cd1284ac81c/external/rules_rust~override~toolchains~rust_linux_x86_64__x86_64-unknown-linux-gnu_tools/BUILD.bazel:89:15: no such package '@[unknown repo 'rules_rust~override~toolchains~rust_linux_x86_64__x86_64-unknown-linux-gnu_tools' requested from @rules_rust~override~toolchains~rust_linux_x86_64__x86_64-unknown-linux-gnu_tools]//': The repository '@[unknown repo 'rules_rust~override~toolchains~rust_linux_x86_64__x86_64-unknown-linux-gnu_tools' requested from @rules_rust~override~toolchains~rust_linux_x86_64__x86_64-unknown-linux-gnu_tools]' could not be resolved: Repository '@rules_rust~override~toolchains~rust_linux_x86_64__x86_64-unknown-linux-gnu_tools' is not visible from repository '@rules_rust~override~toolchains~rust_linux_x86_64__x86_64-unknown-linux-gnu_tools' and referenced by '@rules_rust~override~toolchains~rust_linux_x86_64__x86_64-unknown-linux-gnu_tools//:rust_toolchain'
```

File: rust/repositories.bzl [hunk 2]
```diff
@@ -610,20 +610,20 @@ def rust_repository_set(
...
-
-        if "io_bazel_rules_rust" in native.existing_rules():
-            message += "\n\n" + (
-                "Note that the previous name of \"@io_bazel_rules_rust\" is deprecated.\n" +
-                "See https://github.com/bazelbuild/rules_rust/issues/499 for context."
-            )
-
-        message += "\n" + ("=" * 79)
-        fail(message)
+    # # Inform users that they should be using the canonical name if it's not detected
```
Comment (cameron-martin) [inline rust/repositories.bzl:613]: This errors, because the name of the repository is something like `@rules_rust~override`

**Comment by cameron-martin** [comment]
I've updated the PR with a rough proof of concept that can compile and run a rust binary.

File: rust/repositories.bzl (hunk 2 above)
Comment (meteorcloudy) [reply rust/repositories.bzl:613]: The `native.existing_rules()` API no longer works in Bzlmod. Since we have strict deps in Bzlmod, it's already impossible for the user to depend on something they didn't declare. Maybe we don't need this check anymore?

File: rust/private/repository_utils.bzl (hunk 1 above)
Comment (meteorcloudy) [reply rust/private/repository_utils.bzl:208]: If you want to use the canonical repo name in a label string, you'll need to use `@@canonical_repo_name//foo:bar`, see https://github.com/bazelbuild/bazel/issues/15593

But I guess it's fine to use the label without repo name here since it's referring to targets in the same repo as the BUILD file resides?

Comment (meteorcloudy) [reply rust/private/repository_utils.bzl:208]: /cc @Wyverald

**Comment by meteorcloudy** [comment]
Nice, can you also fix the buildifier errors?

File: examples/bzlmod/hello_world/MODULE.bazel [hunk 3]
```diff
@@ -0,0 +1,18 @@
+module(
+    name = "hello_world",
+    toolchains_to_register = ["@rust_linux_x86_64__x86_64-unknown-linux-gnu//:toolchain"],
```
Comment (meteorcloudy) [inline examples/bzlmod/hello_world/MODULE.bazel:3]: Please use the `register_toolchain` directive instead of the `toolchains_to_register` attribute. Then you can group it together with the `use_rep` directive.

Comment (meteorcloudy) [reply examples/bzlmod/hello_world/MODULE.bazel:3]: See https://github.com/bazelbuild/bazel/issues/15829

File: examples/bzlmod/hello_world/MODULE.bazel [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+)
+
+bazel_dep(name = "rules_rust", version = "0.9.0")
+
+local_path_override(
+    module_name = "rules_rust",
+    path = "../../..",
+)
+
+rust_toolchains = use_extension("@rules_rust//:extensions.bzl", "toolchains")
```
Comment (meteorcloudy) [inline examples/bzlmod/hello_world/MODULE.bazel:14]: I think the toolchain registration can be done in rules_rust's MODULE.bazel file, so that the example module can just depend on rules_rust and get the toolchain for free?

Comment (cameron-martin) [reply examples/bzlmod/hello_world/MODULE.bazel:14]: Which toolchains are available depends on what parameters the user passes to `rust_toolchains.toolchain`. For example, passing `extra_target_triples` generates additional toolchains.

At least, this would be the case if I were to add all the parameters to the tag class that can be forwarded to `rust_register_toolchains`.

File: rust/private/repository_utils.bzl (hunk 1 above)
Comment (cameron-martin) [reply rust/private/repository_utils.bzl:208]: Omitting the repo name looks fine to me in this context.

File: rust/repositories.bzl (hunk 2 above)
Comment (cameron-martin) [reply rust/repositories.bzl:613]: Do we still need this for non-bzlmod usages? Maybe @UebelAndre would know?

If so, I could add a parameter to ignore this check that only is intended to be used by the bzlmod module extension.

File: examples/bzlmod/hello_world/MODULE.bazel (hunk 4 above)
Comment (cameron-martin) [reply examples/bzlmod/hello_world/MODULE.bazel:13]: This does seem to be a significant downside to using bzlmod here. Where previously the user of rules_rust didn't have to know which toolchains were registered, they now have to at the least have the following boilerplate in their MODULE.bazel, as well as know how the options passed to `rust_toolchains.toolchain` map to generated repositories:

```
use_repo(
    rust_toolchains,
    "rust_linux_x86_64__x86_64-unknown-linux-gnu",
    "rust_darwin_aarch64__aarch64-apple-darwin",
    "rust_linux_aarch64__aarch64-unknown-linux-gnu",
    "rust_darwin_x86_64__x86_64-apple-darwin",
    "rust_windows_x86_64__x86_64-pc-windows-msvc",
    "rust_freebsd_x86_64__x86_64-unknown-freebsd",
)

register_toolchains(
    "@rust_linux_x86_64__x86_64-unknown-linux-gnu//:toolchain",
    "@rust_darwin_aarch64__aarch64-apple-darwin//:toolchain",
    "@rust_linux_aarch64__aarch64-unknown-linux-gnu//:toolchain",
    "@rust_darwin_x86_64__x86_64-apple-darwin//:toolchain",
    "@rust_windows_x86_64__x86_64-pc-windows-msvc//:toolchain",
    "@rust_freebsd_x86_64__x86_64-unknown-freebsd//:toolchain",
)
```

I can't think of a way to avoid this currently.

Comment (meteorcloudy) [reply examples/bzlmod/hello_world/MODULE.bazel:13]: It seems we lack of a way to dynamically register generated toolchains. @Wyverald Is it possible to support `register_toolchain` in module extension implementation?

Comment (Wyverald) [reply examples/bzlmod/hello_world/MODULE.bazel:13]: it's possible but we decided not to. I think it's a bit late to introduce it at this point for 6.0. Worst comes to worst, we could have a module called "rules_rust_toolchains" that contains nothing but the above in its MODULE.bazel file.

Comment (illicitonion) [reply examples/bzlmod/hello_world/MODULE.bazel:13]: This may be a pretty significant performance regression - each of these toolchains is in a tar which is hundreds of megabytes, so forcing the user to download 6 of these just to get a toolchain definitions, 5 of which they won't use, is pretty bad.

We could re-structure so that the toolchain definitions live in separate repositories from the actual toolchains, to avoid the proactive downloads, but I wanted to highlight that this is a significant API difference from Bazel.

Comment (Wyverald) [reply examples/bzlmod/hello_world/MODULE.bazel:13]: That's very true. The setup you mentioned is the same as the Java toolchains, so at least there's precedent.

I can think of another potential solution to this. We can make the `toolchains` extension generate a "hub" repo (let's say called "toolchains_hub"), which has nothing but a list of aliases pointing to toolchains that the user requested. And then, we could make rules_rust always register everything in `@toolchains_hub//:all`. That would mean whatever the user requests via tags gets registered, without the user having to do anything. Something like the following:

```bzl
# rules_rust/MODULE.bazel
toolchains = use_extension("//:extensions.bzl", "toolchains")
use_repo(toolchains, "toolchains_hub")
register_toolchains("@toolchains_hub//:all")

# rules_rust/extensions.bzl
toolchains = module_extension(_toolchains_impl, tag_classes = ...)
def _toolchains_impl(ctx):
  toolchain_repos = _somehow_get_repos_from_tags(ctx)
  for toolchain_repo in toolchain_repos: rust_toolchain_repo(**toolchain_repo)
  toolchain_hub_repo(name="toolchains_hub", repos=[tr["name"] for tr in toolchain_repos])
rust_toolchain_repo = repository_rule(...)
toolchain_hub_repo = repository_rule(_toolchain_hub_repo_impl, attrs={"repos": attr.string_list()})
def _toolchain_hub_repo_impl(ctx):
  ctx.file("WORKSPACE")
  ctx.file("BUILD", "\n".join(["alias(name='%s', actual='@%s//:toolchain')" % (repo,repo) for repo in ctx.attr.repos]))

# my_project/MODULE.bazel
bazel_dep(name="rules_rust", version="...")
rust_toolchains = use_extension("//:extensions.bzl", "toolchains")
rust_toolchains.toolchain(edition="2021")
```

Comment (cameron-martin) [reply examples/bzlmod/hello_world/MODULE.bazel:13]: This would work well, as long as using `:all` works correctly for toolchain registration. I'll try it out later.

Comment (cameron-martin) [reply examples/bzlmod/hello_world/MODULE.bazel:13]: Regarding the performance regression, is this any different from the current situation? A lot of unused toolchains get registered currently.

Comment (meteorcloudy) [reply examples/bzlmod/hello_world/MODULE.bazel:13]: Even without Bzlmod, all toolchains get registered in WORKSPACE will be fetched for toolchain resolution, so I think this is the same as the current situation.

Comment (illicitonion) [reply examples/bzlmod/hello_world/MODULE.bazel:13]: > Regarding the performance regression, is this any different from the current situation? A lot of unused toolchains for different platforms get registered currently.

Aah good point, we already split out the meta-toolchain from the expensive repo: https://github.com/bazelbuild/rules_rust/blob/cb9ca1b817febcaf8edc28d160866be759825c31/rust/repositories.bzl#L495-L510

Sorry for the noise!

Comment (cameron-martin) [reply examples/bzlmod/hello_world/MODULE.bazel:13]: It looks like toolchain registration doesn't work with `:all`, since I get the following error when I try to use it:

```
ERROR: /home/cameron/Repos/rules_rust/examples/bzlmod/hello_world/BUILD.bazel:5:12: While resolving toolchains for target //:hello_world: No matching toolchains found for types @rules_rust~override//rust:toolchain_type.
To debug, rerun with --toolchain_resolution_debug='@rules_rust~override//rust:toolchain_type'
```

I've updated the PR to use a hub repo anyway, since it does cut down on the boilerplate.

Comment (Wyverald) [reply examples/bzlmod/hello_world/MODULE.bazel:13]: If registering with `:all` isn't working, it's a bug. We have tests (such as https://cs.opensource.google/bazel/bazel/+/master:src/test/java/com/google/devtools/build/lib/skyframe/ToolchainsForTargetsTest.java;l=219;drc=16c4076cd39f43f3ccabbdeb6ab407c2063b038c) ensuring this behavior.

Do you have more information to help us debug? Check the BUILD file you generated to see whether it actually contains toolchains, etc.?

Comment (cameron-martin) [reply examples/bzlmod/hello_world/MODULE.bazel:13]: The BUILD file contains a bunch of aliases to toolchains:

```
alias(name='rust_darwin_aarch64__aarch64-apple-darwin', actual='@rust_darwin_aarch64__aarch64-apple-darwin//:toolchain', visibility = ["//visibility:public"])
alias(name='rust_linux_aarch64__aarch64-unknown-linux-gnu', actual='@rust_linux_aarch64__aarch64-unknown-linux-gnu//:toolchain', visibility = ["//visibility:public"])
alias(name='rust_darwin_x86_64__x86_64-apple-darwin', actual='@rust_darwin_x86_64__x86_64-apple-darwin//:toolchain', visibility = ["//visibility:public"])
alias(name='rust_windows_x86_64__x86_64-pc-windows-msvc', actual='@rust_windows_x86_64__x86_64-pc-windows-msvc//:toolchain', visibility = ["//visibility:public"])
alias(name='rust_freebsd_x86_64__x86_64-unknown-freebsd', actual='@rust_freebsd_x86_64__x86_64-unknown-freebsd//:toolchain', visibility = ["//visibility:public"])
alias(name='rust_linux_x86_64__x86_64-unknown-linux-gnu', actual='@rust_linux_x86_64__x86_64-unknown-linux-gnu//:toolchain', visibility = ["//visibility:public"])
```

The issue seems to be with the interaction between `:all` and aliases. I've filed a bug report at https://github.com/bazelbuild/bazel/issues/16298.

Comment (cameron-martin) [reply examples/bzlmod/hello_world/MODULE.bazel:13]: I've updated this to put all the toolchain proxies into one repository. This does change the public interface, currently does break non-bzlmod usage and does not include rust-analyzer support, but the idea is there.

File: rust/repositories.bzl [hunk 5]
```diff
@@ -597,7 +608,9 @@ def rust_repository_set(
             sha256s = sha256s,
             urls = urls,
             auth = auth,
-        ))
+        )
+
+        all_toolchain_names.append("@{name}//:toolchain".format(name = name))
 
     # This repository exists to allow `rust_repository_set` to work with the `maybe` wrapper.
```
Comment (cameron-martin) [inline rust/repositories.bzl:615]: I can't say that I really understand the point of this repository. `ALL_TOOLCHAINS` from the repository does not seem to be used anywhere. Can anyone explain this to me?

Comment (UebelAndre) [reply rust/repositories.bzl:615]: To my knowledge, the content of this repository is not used. As the comment states, it's only there to allow `rust_repository_set` to satisfy the `maybe` interface.

Comment (cameron-martin) [reply rust/repositories.bzl:630]: Makes sense.

File: rust/repositories.bzl (hunk 2 above)
Comment (UebelAndre) [reply rust/repositories.bzl:641]: Sorry, missed this one. The code here can be deleted. I opened https://github.com/bazelbuild/rules_rust/pull/1577 to do so.

Comment (cameron-martin) [reply rust/repositories.bzl:641]: Thanks!

**Comment by benbrittain** [comment]
Is there any work I can take on here to help this get landed?

File: extensions.bzl [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+                name = "raw_cargo_{}_{}".format(crate.name, crate.version.replace(".", "_")),
+                sha256 = crate.sha256,
+                url = "https://crates.io/api/v1/crates/{}/{}/download".format(crate.name, crate.version),
+                strip_prefix = "{}-{}".format(crate.name, crate.version),
+                type = "tar.gz",
+                build_file_content = _create_build_file_content(crate.name),
+            )
+
+raw_cargo_crate = tag_class(attrs = {"name": attr.string(), "version": attr.string(), "sha256": attr.string()})
+raw_cargo = module_extension(
```
Comment (jfirebaugh) [inline extensions.bzl:42]: I see `raw_cargo` is used to establish internal dependency on the tinyjson crate. Probably not something intended to be a public API. Perhaps move to `rust/private/extensions.bzl`?

**Comment by UebelAndre** [comment]
@cameron-martin hey, any updates?

**Comment by cameron-martin** [comment]
> @cameron-martin hey, any updates? smile

Sorry I've been busy with other things. I may be able to get back to this in the next week or so.

File: extensions.bzl (hunk 6 above)
Comment (cameron-martin) [reply extensions.bzl:42]: You're right.

**Comment by opicaud** [comment]
hello @cameron-martin, do you have any update ?
Thanks!

**Comment by matts1** [comment]
I've got a prototype working for a module extension for bzlmod. It's still very much a WIP, and needs a lot of cleaning up, but you can at least build examples/bzlmod/crates_repository.

https://github.com/matts1/rules_rust/tree/module_extension_prototype

TLDR on how to use:
```
crates_vendor = use_extension("@rules_rust//:extensions.bzl", "crates_vendor")
crates_vendor.crates_vendor(
    manifests = ["//:Cargo.toml"],
    cargo_lockfile = "//:Cargo.lock",
    lockfile = "//:cargo-bazel-lock.json",
)
use_repo(
    crates_vendor,
    crates = "crates_crates_repository",
)
```

**Comment by cameron-martin** [comment]
I've rebased this PR. Here are the main things that need finishing:

- [ ] Support other types of toolchains (rust analyzer, rustfmt, some wasm stuff that I don't fully understand).
- [ ] Support all options when registering toolchains via bzlmod.
- [ ] Work out how to run the examples without having to cd into examples/bzlmod/hello_world

File: examples/bzlmod/hello_world/MODULE.bazel (hunk 4 above)
Comment (cameron-martin) [reply examples/bzlmod/hello_world/MODULE.bazel:13]: This hub repository now duplicates the toolchain proxies instead of replaces them, so the public API does not change.

**Comment by cameron-martin** [comment]
We also have to consider the API for the module extensions. How many module extensions do we want? Just a single one, called "rust" for everything including crates_universe, etc? One per top-level package, e.g. rust, crates_universe, cargo?

File: MODULE.bazel [hunk 7]
```diff
@@ -0,0 +1,23 @@
+module(
+    name = "rules_rust",
+    version = "0.9.0",
+)
```
Comment (matts1) [inline MODULE.bazel:4]: rules_go put the following in their MODULE.bazel. I think we should add something similar to rules_rust:

`print("WARNING: The rules_go Bazel module is still highly experimental and subject to change at any time. Only use it to try out bzlmod for now.")`

https://github.com/bazelbuild/bazel-central-registry/blob/main/modules/rules_go/0.33.0/MODULE.bazel#L7

**Comment by matts1** [comment]
IMO, we should seperate them into areas of concern, similar to how [other bzlmod modules](https://github.com/bazelbuild/rules_python/blob/main/examples/bzlmod/MODULE.bazel) do it.

File: examples/bzlmod/hello_world/MODULE.bazel [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    version = "1.0",
+)
+
+bazel_dep(name = "rules_rust", version = "0.9.0")
+local_path_override(
+    module_name = "rules_rust",
+    path = "../../..",
+)
+
+rust_toolchains = use_extension("@rules_rust//rust:extensions.bzl", "toolchains")
```
Comment (cameron-martin) [inline examples/bzlmod/hello_world/MODULE.bazel:12]: This module extension should probably be called "rust".

File: MODULE.bazel [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+bazel_dep(name = "platforms", version = "0.0.5")
+bazel_dep(name = "rules_cc", version = "0.0.1")
+bazel_dep(name = "bazel_skylib", version = "1.2.0")
+bazel_dep(name = "apple_support", version = "0.13.0")
+
+# Load tinyjson without using the process wrapper
+raw_cargo = use_extension("//rust/private:extensions.bzl", "raw_cargo")
+raw_cargo.crate(
+    name = "tinyjson",
+    sha256 = "1a8304da9f9370f6a6f9020b7903b044aa9ce3470f300a1fba5bc77c78145a16",
```
Comment (cameron-martin) [inline MODULE.bazel:15]: These values should probably be shared with the non-bzlmod implementation

Comment (matts1) [reply MODULE.bazel:15]: PTAL at the PR I sent you. Amongst other things, it shares the values.

https://github.com/cameron-martin/rules_rust/pull/1/commits/0760cf7516a75eae59dbcf7f20316b6bb55642c6

**Comment by cameron-martin** [comment]
This should now be ready to review.

File: rust/extensions.bzl [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+load(
+    "//rust/private:repository_utils.bzl",
+    "DEFAULT_EXTRA_TARGET_TRIPLES",
+    "DEFAULT_NIGHTLY_VERSION",
+    "DEFAULT_STATIC_RUST_URL_TEMPLATES",
+)
+load(":repositories.bzl", "rust_register_toolchains")
+
+def _rust_impl(ctx):
+    mod = ctx.modules[0]
```
Comment (matts1) [inline rust/extensions.bzl:12]: Nit: Don't think this is guaranteed to be the root module.

`mod = [m for m in ctx.modules if m.is_root][0]`

**Comment by UebelAndre** [comment]
cc @scentini @krasimirgg Hoping one of you two know anything about `bzlmod` and would be able to review this 😅

**Comment by scentini** [comment]
I read up a bit on `bzlmod` and unless someone with in-depth knowledge offers to do the review, I'll claim the 'expert' title. This PR LGTM, and aside from rebasing and updating the version in `MODULE.bazel`, I'd try merging.

File: MODULE.bazel [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+print("WARNING: The rules_rust Bazel module is still highly experimental and subject to change at any time. Only use it to try out bzlmod for now.")  # buildifier: disable=print
+
+bazel_dep(name = "platforms", version = "0.0.5")
+bazel_dep(name = "rules_cc", version = "0.0.1")
+bazel_dep(name = "bazel_skylib", version = "1.2.0")
+bazel_dep(name = "apple_support", version = "1.3.1")
+
+non_bzlmod_deps = use_extension("//rust/private:extensions.bzl", "non_bzlmod_deps")
+non_bzlmod_deps.non_bzlmod_deps()
```
Comment (Wyverald) [inline MODULE.bazel:14]: this line is unnecessary. the `use_extension` clause is enough to triger the extension

**Review by Wyverald — COMMENTED** [review]
had a brief pass, mostly looks good! I didn't read the actual implementation very deeply.

File: MODULE.bazel [hunk 12]
```diff
@@ -0,0 +1,18 @@
+module(
+    name = "rules_rust",
+    version = "0.9.0",
```
[the remark this answers is not in the document]
Comment (cameron-martin) [reply MODULE.bazel]: I've updated this to the same version as in `version.bzl`

File: rust/private/extensions.bzl [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+
+def _non_bzlmod_deps_impl(_module_ctx):
+    http_archive(**TINYJSON_KWARGS)
+
+non_bzlmod_deps_non_bzlmod_deps = tag_class(attrs = {})
+non_bzlmod_deps = module_extension(
+    doc = "Dependencies for rules_rust",
+    implementation = _non_bzlmod_deps_impl,
+    tag_classes = dict(
+        non_bzlmod_deps = non_bzlmod_deps_non_bzlmod_deps,
```
[the remark this answers is not in the document]
Comment (cameron-martin) [reply rust/private/extensions.bzl]: Renamed to `internal_deps`

File: MODULE.bazel (hunk 11 above)
Comment (cameron-martin) [reply MODULE.bazel]: Fixed

**Comment by scentini** [comment]
Thank you @Wyverald !

**Review by scentini — APPROVED** [review]

**Comment by meteorcloudy** [comment]
Nice! Will there be the next step to integrate with cargo to actually fetch rust dependencies?

File: examples/bzlmod/hello_world/MODULE.bazel [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+    version = "1.0",
+)
+
+bazel_dep(name = "rules_rust", version = "0.9.0")
+local_path_override(
+    module_name = "rules_rust",
+    path = "../../..",
+)
+
+rust = use_extension("@rules_rust//rust:extensions.bzl", "rust")
```
[the remark this answers is not in the document]
Comment (cameron-martin) [reply examples/bzlmod/hello_world/MODULE.bazel:12]: Currently only the toolchains declared in the top level module are registered. There was an unresolved discussion on whether this was the correct thing to do [here](https://github.com/bazelbuild/rules_rust/pull/1842#discussion_r1116293025), but currently the toolchain that rules_rust defines would not get registered.

**Comment by cameron-martin** [comment]
> Nice! Will there be the next step to integrate with cargo to actually fetch rust dependencies?

@matts1 currently has a PR for this here: https://github.com/bazelbuild/rules_rust/pull/1910

File: examples/bzlmod/hello_world/MODULE.bazel (hunk 14 above)
[the remark this answers is not in the document]
Comment (meteorcloudy) [reply examples/bzlmod/hello_world/MODULE.bazel:12]: Got it, thanks!
